### PR TITLE
Increased configuration parameters to match with increased support in options.json

### PR DESCRIPTION
**Pull request by GeoDerp**
Based from: https://github.com/davidusb-geek/emhass/pull/149

### CODE REVIEW

**Comment by GeoDerp** [comment]
The thought was that items (such as `solcast_api_key`) commonly used in curl post commands could be redacted from this poll request.

The reasoning for adding all parameters into the configuration page was to assist in new EMHASS users. 
All parameters in one location may help users to understand what parameters could be changed, while placing them all in a central location. This also results in an minimal integration of the automations.yaml & configuration.yaml. The thought is that most people could use the base examples:

>  In configuration.yaml:
> 
> shell_command:
>   dayahead_optim: "curl -i -H \"Content-Type:application/json\" -X POST -d '{}' http://localhost:5000/action/dayahead-optim"
>   publish_data: "curl -i -H \"Content-Type:application/json\" -X POST -d '{}' http://localhost:5000/action/publish-data"
> 
> In automations.yaml:
> 
> - alias: EMHASS day-ahead optimization
>   trigger:
>     platform: time
>     at: '05:30:00'
>   action:
>   - service: shell_command.dayahead_optim
> - alias: EMHASS publish data
>   trigger:
>   - minutes: /5
>     platform: time_pattern
>   action:
>   - service: shell_command.publish_data

In most cases to develop a functioning EMHASS setup.
Some examples of automating (with bash commands & the terminal addon for example) could be documented later for an copy and past commands of adding the additions into there current setup.

**Comment by GeoDerp** [comment]
The downside to this method is that it will increase the configuration page bloat https://github.com/davidusb-geek/emhass-add-on/issues/48. Im hopping to see or request some way for home assistant configuration pages to support dropdowns or sections to make the parameters easier for the viewer

**Comment by GeoDerp** [comment]
Update: ill found the method of hiding optional  parameters. Will look into adding later

**Comment by GeoDerp** [comment]
This currently does not work. I believe it may be to do with the optional parameters feeding params with values into EMHASS overriding the cones provided in config_emhass.yaml

**Comment by GeoDerp** [comment]
Commit https://github.com/davidusb-geek/emhass-add-on/pull/69/commits/63f24838347a0f455ae29955ac171efbb303ea01  currently does not work. I believe it may be from the optional parameters feeding empty values into EMHASS, overriding the the param values provided in config_emhass.yaml

**Comment by GeoDerp** [comment]
your choice, my idea of fixing this could be something like  
 
https://github.com/davidusb-geek/emhass/blob/b66065e4fec74a1d400c0835f812d8539565b2d6/src/emhass/web_server.py#L100-L164
  
to
  
  
```python
def build_params(params, options, addon):
    if addon == 1:
        #create associations list
        # conf, key, options name, 'param in list name (for lists)' 
        associations = []    
        # variables in retrieve_hass_conf
        associations.append['retrieve_hass_conf', 'freq', 'optimization_time_step']
        associations.append['retrieve_hass_conf', 'days_to_retrieve', 'historic_days_to_retrieve']
        associations.append['retrieve_hass_conf', 'var_PV', 'sensor_power_photovoltaics']
        associations.append['retrieve_hass_conf', 'var_load', 'sensor_power_load_no_var_loads']
        associations.append['retrieve_hass_conf', 'load_negative', 'load_negative']
        associations.append['retrieve_hass_conf', 'set_zero_min', 'set_zero_min']
        associations.append['retrieve_hass_conf', 'var_replace_zero', 'sensor_power_photovoltaics']
        associations.append['retrieve_hass_conf', 'method_ts_round', 'method_ts_round']
        associations.append['retrieve_hass_conf', 'solcast_api_key', 'optional_solcast_api_key']
        associations.append['retrieve_hass_conf', 'solcast_rooftop_id', 'optional_solcast_rooftop_id']
        associations.append['retrieve_hass_conf', 'solar_forecast_kwp', 'optional_solar_forecast_kwp']
        associations.append['retrieve_hass_conf', 'time_zone', 'time_zone']
        associations.append['retrieve_hass_conf', 'lat', 'Latitude']
        associations.append['retrieve_hass_conf', 'lon', 'Longitude']
        associations.append['retrieve_hass_conf', 'alt', 'Altitude']
        #  variables in optim_conf
        associations.append['optim_conf', 'set_use_battery', 'set_use_battery']
        associations.append['optim_conf', 'num_def_loads', 'number_of_deferrable_loads']
        associations.append['optim_conf', 'P_deferrable_nom', 'list_nominal_power_of_deferrable_loads', 'nominal_power_of_deferrable_loads']
        associations.append['optim_conf', 'def_total_hours', 'list_operating_hours_of_each_deferrable_load', 'operating_hours_of_each_deferrable_load']
        associations.append['optim_conf', 'treat_def_as_semi_cont', 'list_treat_deferrable_load_as_semi_cont', 'treat_deferrable_load_as_semi_cont']
        associations.append['optim_conf', 'set_def_constant', 'list_set_deferrable_load_single_constant', 'set_deferrable_load_single_constant']
        associations.append['optim_conf', 'weather_forecast_method', 'weather_forecast_method']
        associations.append['optim_conf', 'load_forecast_method', 'load_forecast_method']
        associations.append['optim_conf', 'delta_forecast', 'delta_forecast_daily']
        associations.append['optim_conf', 'load_cost_forecast_method', 'load_cost_forecast_method']
        associations.append['optim_conf', 'load_cost_hp', 'load_peak_hours_cost']
        associations.append['optim_conf', 'load_cost_hc', 'load_offpeak_hours_cost']
        associations.append['optim_conf', 'prod_price_forecast_method', 'production_price_forecast_method']
        associations.append['optim_conf', 'prod_sell_price', 'photovoltaic_production_sell_price']
        associations.append['optim_conf', 'set_total_pv_sell', 'set_total_pv_sell']
        associations.append['optim_conf', 'lp_solver', 'lp_solver']
        associations.append['optim_conf', 'lp_solver_path', 'lp_solver_path']
        associations.append['optim_conf', 'set_nocharge_from_grid', 'set_nocharge_from_grid']
        associations.append['optim_conf', 'set_nodischarge_to_grid', 'set_nodischarge_to_grid']
        associations.append['optim_conf', 'set_battery_dynamic', 'set_battery_dynamic']
        associations.append['optim_conf', 'battery_dynamic_max', 'battery_dynamic_max']
        associations.append['optim_conf', 'battery_dynamic_min', 'battery_dynamic_min']
        associations.append['optim_conf', 'weight_battery_discharge', 'weight_battery_discharge']
        associations.append['optim_conf', 'weight_battery_charge', 'weight_battery_charge']
        # variables in plant_conf
        associations.append['plant_conf', 'P_grid_max', 'maximum_power_from_grid']
        associations.append['plant_conf', 'module_model', 'list_pv_module_model', 'pv_module_model' ]
        associations.append['plant_conf', 'inverter_model', 'list_pv_inverter_model', 'pv_inverter_model' ]
        associations.append['plant_conf', 'surface_tilt', 'list_surface_tilt', 'surface_tilt' ]
        associations.append['plant_conf', 'surface_azimuth', 'list_surface_azimuth', 'surface_azimuth']
        associations.append['plant_conf', 'modules_per_string','list_modules_per_string', 'modules_per_string']
        associations.append['plant_conf', 'strings_per_inverter', 'list_strings_per_inverter', 'strings_per_inverter']
        associations.append['plant_conf', 'Pd_max', 'battery_discharge_power_max']
        associations.append['plant_conf', 'Pc_max', 'battery_charge_power_max']
        associations.append['plant_conf', 'eta_disch', 'battery_discharge_efficiency']
        associations.append['plant_conf', 'eta_ch', 'battery_charge_efficiency']
        associations.append['plant_conf', 'Enom', 'battery_nominal_energy_capacity']
        associations.append['plant_conf', 'SOCmin', 'battery_minimum_state_of_charge']
        associations.append['plant_conf', 'SOCmax', 'battery_maximum_state_of_charge']
        associations.append['plant_conf', 'SOCtarget', 'battery_target_state_of_charge']
        
        print(associations)

        for i in associations:
            # check if options are null 
            if  options[str(i[2])] is not None:
                #check if list
                if options[str(i[3])] is not None:
                    params[str(i[0])][str(i[1])] = [x[str(i[3])] for x in options[str(i[2])]]
                #check if dictionary
                elif type(options[str(i[2])]) is not dict:
                    params[str(i[0])][str(i[1])] = options[str(i[2])]
                else:    
                    params[str(i[0])][str(i[1])] = [options[str(i[2])]]

        ## add rest manually 
        params['retrieve_hass_conf']['var_interp'] = [options['sensor_power_photovoltaics'], options['sensor_power_load_no_var_loads']]
        start_hours_list = [i['peak_hours_periods_start_hours'] for i in options['list_peak_hours_periods_start_hours']]
        end_hours_list = [i['peak_hours_periods_end_hours'] for i in options['list_peak_hours_periods_end_hours']]
        num_peak_hours = len(start_hours_list)
        list_hp_periods_list = [{'period_hp_'+str(i+1):[{'start':start_hours_list[i]},{'end':end_hours_list[i]}]} for i in range(num_peak_hours)]
        params['optim_conf']['list_hp_periods'] = list_hp_periods_list

        print(params)
```

**Comment by GeoDerp** [comment]
e616c364d307360bb1a428a4415cc094eec78c54

**Comment by GeoDerp** [comment]
Sorry about that. I tested it via python and Docker. Seems like my testing still needs improvement.

> Did you meant?:
> ```
> associations = []    
> associations.append(['retrieve_hass_conf', 'freq', 'optimization_time_step'])
> associations.append(['retrieve_hass_conf', 'days_to_retrieve', 'historic_days_to_retrieve'])
> ```
> ...

I'm not surprised that I got a few parameters wrong. Will change and test Tomorrow. 👍
Let me know if there are any tests besides the pytest and running the docker container you would recommend for verifying.

**Comment by GeoDerp** [comment]
Another idea *(besides idea: https://github.com/davidusb-geek/emhass/compare/master...GeoDerp:emhass:testing )* I had was setting get methods with each option. 

ex: *(web_server/py)*
```python
 params['retrieve_hass_conf']['freq'] = options['optimization_time_step']
 ```
 to 
 ```python
  params['retrieve_hass_conf']['freq'] = options.get('optimization_time_step', 30)
 ``` 
 
If it did work, would require setting defaults on all values.

Unfortunately i have hit an wall. I am unable to continue due to me lack of knowledge of how to properly locally build emhass to  test emhass-add-on. Any help would be appreciated.

Thanks for putting up with me this far.

**Comment by davidusb-geek** [comment]
@GeoDerp let's try to come out with a new release that includes all your changes until now. Then if you have any more ideas you can open new PR's.

Yes sometime ago it was easy for me to test the add-on by using this procedure locally: https://developers.home-assistant.io/docs/add-ons/testing

This was very useful to test the add-on on a production-like environment.
However, several month ago this stopped working for me as the devcontainer always fail to start the test Home Assistant instance with errors. 

So what I have done lately is to build the docker standalone image and locally running it to test the web server this way.

**Review by davidusb-geek — APPROVED** [review]

File: emhass/config.yml [hunk 1]
```diff
@@ -24,24 +24,15 @@ ingress: true
 ingress_port: 5000
 ingress_stream: true
 options:
-  hass_url: empty
```
[the remark this answers is not in the document]
Comment (GeoDerp) [reply emhass/config.yml:27]: yep, its a weird way to do it. but thats how Home Assistant wants it

**Pull request merged**